pkg/volume/util/operationexecutor/operation_executor_test.go

```diff
@@ -0,0 +1,201 @@
+	"k8s.io/kubernetes/pkg/api/v1"
+	"k8s.io/kubernetes/pkg/types"
+	"k8s.io/kubernetes/pkg/util/mount"
+	"strconv"
+	"testing"
+	"time"
+)
+
+var _ OperationGenerator = &mockOperationGenerator{}
+
+func TestOperationExecutor_MountVolume_ParallelMountForNonAttachablePlugins(t *testing.T) {
+	// Arrange
+	oe := NewOperationExecutor(newMockOperationGenerator())
+
+	numVolumesToMount := 2
+	volumesToMount := make([]VolumeToMount, numVolumesToMount)
+	secretName := "secret-volume"
+	volumeName := v1.UniqueVolumeName(secretName)
+
+	// Act
+	for i := range volumesToMount {
+		podName := "pod-" + strconv.Itoa((i + 1))
+		pod := getTestPodWithSecret(podName, secretName)
+		volumesToMount[i] = VolumeToMount{
+			Pod:                pod,
+			VolumeName:         volumeName,
+			PluginIsAttachable: false, /*this field determines whether the plugin is attachable*/
+			ReportedInUse:      true,
+		}
+		oe.MountVolume(0 /*waitForAttachTimeOut*/, volumesToMount[i], nil /*actualStateOfWorldMounterUpdater*/)
+	}
+	// This sleep is to ensure that we don't assert before the mount operations have made it to
+	// the operations data structure
+	time.Sleep(1 * time.Second)
+
+	// Assert
+	if len(oe.GetNestedPendingOperations().GetOperations()) != numVolumesToMount {
+		t.Fatalf("Unable to mount in parallel for a non-attachable plugin")
+	}
+}
+
+func TestOperationExecutor_MountVolume_ParallelMountForAttachablePlugins(t *testing.T) {
+	// Arrange
+	oe := NewOperationExecutor(newMockOperationGenerator())
+
+	numVolumesToMount := 2
+	volumesToMount := make([]VolumeToMount, numVolumesToMount)
+	pdName := "pd-volume"
+	volumeName := v1.UniqueVolumeName(pdName)
+
+	// Act
+	for i := range volumesToMount {
+		podName := "pod-" + string((i + 1))
+		pod := getTestPodWithGCEPD(podName, pdName)
+		volumesToMount[i] = VolumeToMount{
+			Pod:                pod,
+			VolumeName:         volumeName,
+			PluginIsAttachable: true, /*this field determines whether the plugin is attachable*/
+			ReportedInUse:      true,
+		}
+		oe.MountVolume(0 /*waitForAttachTimeout*/, volumesToMount[i], nil /*actualStateOfWorldMounterUpdater*/)
+	}
+	// This sleep is to ensure that we don't assert before the mount operations have made it to
+	// the operations data structure
+	time.Sleep(1 * time.Second)
+
+	// Assert
+	if len(oe.GetNestedPendingOperations().GetOperations()) != 1 {
+		t.Fatalf("Parallel mount shouldn't happen for attachable plugins")
+	}
+}
+
+type mockOperationGenerator struct {
+}
+
+func newMockOperationGenerator() OperationGenerator {
+	return &mockOperationGenerator{}
+}
+
+func (mopg *mockOperationGenerator) GenerateMountVolumeFunc(waitForAttachTimeout time.Duration, volumeToMount VolumeToMount, actualStateOfWorldMounterUpdater ActualStateOfWorldMounterUpdater) (func() error, error) {
+	return func() error {
+		// Mocking the behavior of mount to take a long time to
+		// assert if mounts can be done in parallel
+		time.Sleep(10 * time.Second)
```

Comment:
Relying on time will cause these test to be flaky. Instead you should use channels to synchronize their behavior. Basically have this method block until something is passed down that channel. And your test will only pass something down the channel after it is done verifying that things are actually running in parallel. See https://github.com/kubernetes/kubernetes/blob/master/pkg/util/goroutinemap/goroutinemap_test.go for examples of this.